lib/http_aws_sigv4.c

```diff
@@ -66,6 +68,159 @@ static void sha256_to_hex(char *dst, unsigned char *sha, size_t dst_l)
   }
 }
 
+
+
+static CURLcode make_headers(struct Curl_easy *data,
+                             const char *hostname,
+                             const char *timestamp,
+                             const char *provider1_low,
+                             const char *provider1_mid,
+                             char **canonical_headers,
+                             char **signed_headers)
+{
+  char *date_hdr_key =  curl_maprintf("X-%s-Date", provider1_mid);
+  char *date_full_hdr = curl_maprintf("x-%s-date:%s",
+                                      provider1_low, timestamp);
+#ifdef DEBUGBUILD
+  char *force_host = getenv("CURL_FORCEHOST");
+#endif
+  struct curl_slist *head = NULL;
+  struct curl_slist *tmp_head = NULL;
+  int ret = CURLE_OUT_OF_MEMORY;
+  struct curl_slist *l;
+  int again = 1;
+
+  if(!date_hdr_key || !date_full_hdr)
+    goto fail;
+
+  if(Curl_checkheaders(data, "Host")) {
+    head = NULL;
+  }
+#ifdef DEBUGBUILD
+  else if(force_host) {
+    head = curl_slist_append(NULL, "host:127.0.0.1");
+    if(!head)
+      goto fail;
+  }
+#endif
+  else {
+    char *full_host = data->state.aptr.host ? data->state.aptr.host :
+      curl_maprintf("host:%s", hostname);
+
+    if(!full_host)
+      goto fail;
+    head = curl_slist_append(NULL, full_host);
+    if(full_host != data->state.aptr.host)
+      free(full_host);
+    if(!head)
+      goto fail;
+  }
+
+
+  if(!date_hdr_key || !date_full_hdr)
+    goto fail;
+
+  for(l = data->set.headers; l; l = l->next) {
+    tmp_head = curl_slist_append(head, l->data);
+    if(!tmp_head)
+      goto fail;
+    head = tmp_head;
+  }
+
+  /* remove whitespace, and lowercase each headers */
+  for(l = head; l; l = l->next) {
+    char *tmp = l->data, *tmp2;
+
+    Curl_strntolower(l->data, l->data, strlen(l->data));
```

Comment:
Sorry to bother. 

Here you lowercase the whole header while the specs require you to lowercase only the header name:

https://docs.aws.amazon.com/AmazonS3/latest/API/sig-v4-header-based-auth.html

Reply:
Thanks, now the headers are lower-cased properly.

Reply:
Thanks for the feedback :)